src/Security/Authentication/Negotiate/src/Internal/ReflectedNegotiateState.cs

```diff
@@ -68,9 +94,65 @@ public string GetOutgoingBlob(string incomingBlob)
             return outgoingBlob;
         }
 
-        private byte[] GetOutgoingBlob(byte[] incomingBlob, bool thrownOnError)
+        private byte[] GetOutgoingBlob(byte[] incomingBlob, out BlobErrorType status, out Exception error)
         {
-            return (byte[])_getOutgoingBlob.Invoke(_instance, new object[] { incomingBlob, thrownOnError });
+            try
+            {
+                // byte[] GetOutgoingBlob(byte[] incomingBlob, bool throwOnError, out SecurityStatusPal statusCode)
+                var parameters = new object[] { incomingBlob, false, null };
+                var blob = (byte[])_getOutgoingBlob.Invoke(_instance, parameters);
+
+                var securityStatus = parameters[2];
+                // TODO: Update after corefx changes
+                error = (Exception)(_statusException.GetValue(securityStatus)
+                    ?? _getException.Invoke(null, new[] { securityStatus }));
+                var errorCode = (SecurityStatusPalErrorCode)_statusCode.GetValue(securityStatus);
+
+                // TODO: Remove after corefx changes
+                // The linux implementation always uses InternalError;
+                if (errorCode == SecurityStatusPalErrorCode.InternalError
+                    && !RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                    && _gssExceptionType.IsInstanceOfType(error))
+                {
+                    var majorStatus = (uint)error.HResult;
+                    var minorStatus = (uint)_gssMinorStatus.GetValue(error);
+
+                    // Remap specific errors
+                    if (majorStatus == GSS_S_NO_CRED && minorStatus == 0)
+                    {
+                        errorCode = SecurityStatusPalErrorCode.UnknownCredentials;
+                    }
+
+                    error = new Exception($"An authentication exception occured (0x{majorStatus:X}/0x{minorStatus:X}).", error);
+                }
+
+                if (errorCode == SecurityStatusPalErrorCode.OK
+                    || errorCode == SecurityStatusPalErrorCode.ContinueNeeded
+                    || errorCode == SecurityStatusPalErrorCode.CompleteNeeded)
+                {
+                    status = BlobErrorType.None;
+                }
+                else if (IsCredentailError(errorCode))
+                {
+                    status = BlobErrorType.CredentialError;
+                }
+                else if (IsClientError(errorCode))
+                {
+                    status = BlobErrorType.ClientError;
+                }
+                else
+                {
+                    status = BlobErrorType.Other;
+                }
+
+                return blob;
+            }
+            catch (TargetInvocationException tex)
+            {
+                // Unwrap
+                ExceptionDispatchInfo.Capture(tex.InnerException).Throw();
+                throw;
```

Comment:
I don't think you need the second throw right?  Capture.Throw() should throw?

Reply:
Capture.Throw() does throw but the compiler doesn't know that, it requires a `throw` or a `return` on all code paths. This line will never be hit though.